third_party/move/tools/move-package-cache/src/canonical.rs

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use move_core_types::account_address::AccountAddress;
+use url::Url;
+
+/// Sanitizes a URL by replacing dots with underscores and converting to lowercase.
+///
+/// TODO: The current scheme is not bijective - if the original URL contains underscores, they will
+/// be preserved while dots are converted to underscores, creating ambiguity.
+fn sanitized_url(url: &Url) -> String {
+    url.host_str()
+        .unwrap_or("unknown")
+        .replace('.', "_")
+        .to_ascii_lowercase()
+}
+
+/// Creates a canonical name for a git repository from its URL.
+/// This is used to identify and index the repo in the package cache.
+///
+/// # Example
+/// `https://github.com/aptos-labs/aptos-core` -> `github_com+aptos-labs+aptos-core`
+pub fn canonical_repo_name_from_url(url: &Url) -> String {
+    let mut s = String::new();
+    s.push_str(&sanitized_url(url));
+    if let Some(segments) = url.path_segments() {
+        for segment in segments {
+            s.push('+');
+            s.push_str(&segment.replace('.', "_").to_ascii_lowercase());
+        }
+    }
+    s
+}
+
+/// Creates a canonical name for an on-chain package, used to identify and index the package.
+///
+/// # Example
+/// `fullnode.mainnet.aptoslabs.com, version 12345, 0x1::std`
+///   -> `fullnode_mainnet_aptoslabs_com+12345+0x1+std`
```

Comment:
Again, not quite sure why not keeping domain name syntax. For parameters (not paths), `+` is though appropriate.

If you are looking to map this to file names, perhaps the mangling should be done in the file backend.

Reply:
Same, this is used as a directory name. 

> perhaps the mangling should be done in the file backend

Could you clarify what you meant by this?